concrete-core/docs/api/backends.md

```diff
@@ -0,0 +1,243 @@
++ Uniform binary
++ Uniform ternary
++ Gaussian
+
+The distribution of the key used to encrypt a ciphertext has a direct impact on which algorithm can be used
+on the ciphertext. For this reason, we have a mechanism to ensure, at compile time, that algorithms gets
+executed on compatible entities.
+
+This mechanism relies on associating a marker type to keys and ciphertexts. For instance, the
+`LweSecretKeyEntity` has an associated `KeyDistribution` type in its definition, which represent which
+sampling was used:
+```rust
+pub trait LweSecretKeyEntity: AbstractEntity<Kind = LweSecretKeyKind> {
+    type KeyDistribution: KeyDistributionMarker;
+    fn lwe_dimension(&self) -> LweDimension;
+}
+```
+
+This `KeyDistribution` type must implement `KeyDistributionMarker`, which means it can be one of three types:
++ `BinaryKeyDistribution`
++ `TernaryKeyDistribution`
++ `GaussianKeyDistribution`
+
+Those types are empty and have no other use than representing the distribution of the key in the type system.
+
+The `backends::core::entities::LweSecretKey64` type implements `LweSecretKeyEntity` with the `KeyDistribution`
+type equal to `BinaryKeyDistribution` (see
+[this line of the doc](https://docs.rs/concrete-core/latest/concrete_core/backends/core/entities/struct.LweSecretKey64.html#associatedtype.KeyDistribution)).
+This means that it represents binary lwe secret keys.
+
+Now, if we have a look at the definition of the `LweCiphertextEntity` trait:
+```rust
+pub trait LweCiphertextEntity: AbstractEntity<Kind = LweCiphertextKind> {
+    type KeyDistribution: KeyDistributionMarker;
+    fn lwe_dimension(&self) -> LweDimension;
+}
+```
+
+We can see that it also contains a `KeyDistribution` associated type, which represents which kind of key was
+used when encrypting the ciphertext.
+
+The `backends::core::entities::LweCiphertext64` type implements `LweCiphertextEntity` with the
+`KeyDistribution` type also equals to `BinaryKeyDistribution` (see
+[this line of the doc](https://docs.rs/concrete-core/latest/concrete_core/backends/core/entities/struct.LweCiphertext64.html#associatedtype.KeyDistribution)).
+This means that it represents an lwe ciphertext encoded with a binary key.
+
+Most importantly, the definition of `LweCiphertextEncryptionEngine` looks like this:
+```rust
+pub trait LweCiphertextEncryptionEngine<SecretKey, Plaintext, Ciphertext>: AbstractEngine where
+    SecretKey: LweSecretKeyEntity,
+    Plaintext: PlaintextEntity,
+    Ciphertext: LweCiphertextEntity<KeyDistribution = SecretKey::KeyDistribution>, {
+    ...
+}
+```
+
+If you focus on the trait bounds, you can see
+`Ciphertext: LweCiphertextEntity<KeyDistribution = SecretKey::KeyDistribution>`. This bounds prevents
+implementing `LweCiphertextEncryptionEngine` for two generic types `SecretKey` and `Ciphertext` with
+incompatible key distribution.
+
+For the `backends::core::engines::CoreEngine` type implements `LweCiphertextEncryptionEngine` for
```

Comment:
not sure about keeping the leading `For the` or maybe I dont understand this sentence